src/napari_matplotlib/histogram.py

```diff
@@ -29,26 +39,143 @@ def __init__(
-        """
-        Clear the axes and histogram the currently selected layer/slice.
-        """
-        layer = self.layers[0]
+    @property
+    def bins_start(self) -> float:
+        """Minimum bin edge"""
+        return self.findChild(QDoubleSpinBox, name="bins start").value()
 
+    @bins_start.setter
+    def bins_start(self, start: Union[int, float]) -> None:
+        """Set the minimum bin edge"""
+        self.findChild(QDoubleSpinBox, name="bins start").setValue(start)
+
+    @property
+    def bins_stop(self) -> float:
+        """Maximum bin edge"""
+        return self.findChild(QDoubleSpinBox, name="bins stop").value()
+
+    @bins_stop.setter
+    def bins_stop(self, stop: Union[int, float]) -> None:
+        """Set the maximum bin edge"""
+        self.findChild(QDoubleSpinBox, name="bins stop").setValue(stop)
+
+    @property
+    def bins_num(self) -> int:
+        """Number of bins to use"""
+        return self.findChild(QSpinBox, name="bins num").value()
+
+    @bins_num.setter
+    def bins_num(self, num: int) -> None:
+        """Set the number of bins to use"""
+        self.findChild(QSpinBox, name="bins num").setValue(num)
+
+    def autoset_widget_bins(self, data: npt.NDArray[Any]) -> None:
+        """Update widgets with bins determined from the image data"""
+        bins = np.linspace(np.min(data), np.max(data), 100, dtype=data.dtype)
+        self.bins_start = bins[0]
+        self.bins_stop = bins[-1]
+        self.bins_num = bins.size
+
+    def _get_layer_data(self, layer: napari.layers.Layer) -> npt.NDArray[Any]:
+        """Get the data associated with a given layer"""
         if layer.data.ndim - layer.rgb == 3:
             # 3D data, can be single channel or RGB
             data = layer.data[self.current_z]
             self.axes.set_title(f"z={self.current_z}")
         else:
             data = layer.data
+
         # Read data into memory if it's a dask array
         data = np.asarray(data)
 
+        return data
+
+    def on_update_layers(self) -> None:
+        """
+        Called when the layer selection changes by ``self._update_layers()``.
+        """
+        if not self.layers:
+            return
+
+        # Reset to bin start, stop and step
+        layer_data = self._get_layer_data(self.layers[0])
+        self.autoset_widget_bins(data=layer_data)
+
+        # Only allow integer bins for integer data
+        # And only allow values greater than 0 for unsigned integers
+        n_decimals = 0 if np.issubdtype(layer_data.dtype, np.integer) else 2
+        is_unsigned = layer_data.dtype.kind == "u"
+        minimum_value = 0 if is_unsigned else -1e10
+
+        bins_start = self.findChild(QDoubleSpinBox, name="bins start")
+        bins_stop = self.findChild(QDoubleSpinBox, name="bins stop")
+        bins_start.setDecimals(n_decimals)
+        bins_stop.setDecimals(n_decimals)
+        bins_start.setMinimum(minimum_value)
+        bins_stop.setMinimum(minimum_value)
```

Comment:
it's necessary to set the minimum bin values to 0 for unsigned integers - if negative bin limits are selected for data with unsigned integers, `plt.hist` complains that bins must be monotonically increasing (as there's an integer underflow in the bins produced by`np.linspace`)

Reply:
it's not necessary to set the decimals to 0 for integers, but makes it clear to the users that the data are integers and float bin limits are not appropriate